Add support for reading >2 GiB files from disk

augustg opened the pull request:
### Background
When I download a new song pack I always put it through a short process of removing the "+9ms bias" making sure the songs are null synced. I then re-package the songs as a package ("smzip") which I copy to the Packages/ folder.
There is no other benefit to this than me thinking it is a neat way of organizing song packs and being able to label each song pack as "null synced" in the filename, without having to change the name of the song group folder.

### The problem
ITGmania (and StepMania) cannot read files larger than 2 GiB due to using signed 32-bit integers to seek in files.
I discovered this after making a package out of Stamina RPG 6, which was 3.6 GiB.

### Changes
- Read >2 GiB files from disk (for our purposes the file size is now _practically_ unlimited).
- Read up to 4 GiB zip files, up from 2 GiB.
- Read up to 4 GiB files from within a zip file, up from 2 GiB.

The reason why only 4 GiB is supported for zip files is that I have only corrected the simple mistake of using a signed 32-bit integer to represent the sizes of files within zip files, and instead using unsigned 32-bit integer.
To support larger zip files we would have to implement ZIP64, which is a bigger effort and out of scope for this PR.
Given that the community has made do with a 2 GiB limit for 20 years, 4 GiB should last us another 20.

### Testing
I've tested this on windows only. I currently dont have a linux dev env setup, but I do run ubuntu on my pad setup (smx pad).
Given the foundational nature of these changes there may be edge cases that are hard to find.

augustg committed:
Add support for reading >2 GiB files from disk

- Read >2 GiB files from disk.
- Read up to 4 GiB zip files, up from 2 GiB.
- Read up to 4 GiB files from within a zip file, up from 2 GiB.

geefr commented:
At a glance - looks fine, but yeah should be tested extensively.
In theory it's a straightforward one though, since the underlying file apis are all 64-bit anyway?
(Or are there some older win32 apis being used that are 32-bit?)

Are there parts that interact with external file apis, e.g. passing a file offset back/forth to ffmpeg, through some 32-bit api?

If these changes are being made, is it also a good time to delete RageFile? I'm assuming not (since RageFile also accesses some virtual file system for config, logs, etc. It's worth asking "should we delete Rage X" anyways though)

FishamanP commented:
The song pack Speirmix Galaxy is distributed as a 6GB ZIP64. I think it would be worth tackling the required work in the same PR, although I'm not a contributor.

augustg commented:
@geefr As far as I saw, only the lseek api had to be replaced with its 64-bit equivalent. We now use lseek64.
When reading data from a file, 32-bit buffer sizes are not a problem.

Is there an ongoing effort to replace RageFile? It seemed to be quite heavily in use.

@FishamanP Yes, ZIP64 would be nice. If I feel like doing that in the future, I'd like to do it in a separate PR.

geefr commented:
Gotcha, I guess other than the testing point then (linux & mac) I'd say this looks good

I doubt anyone's brave enough to rework Rage itself, past small fixes here and there..Just a blue sky thought worth asking, but in reality if it works it's probably best not to touch it.

teejusb commented:
I can test with Linux and Mac I think. Do you have a > 2GB pack I can test with? Also is the procedure just to run the game and see if it loads up correctly?

I do think upgrading to 64bit ZIP64 sounds better at some point but this seems fine as a middling solution.

augustg committed:
Fix linux compile errors

augustg commented:
[Heres an .smzip](https://drive.google.com/file/d/1gAN6RzO5TWpCHACX5koHCMPS2dJfJqSV/view?usp=sharing) that is both larger than 2 GiB and contains a background video file larger than 2 GiB.

1. Put the zip in itgmania/Packages.
2. If the game starts, the zip file was successfully mounted (not working on master branch).
3. If the background video plays during the song included in the zip large files can also successfully be read from within zip files (also not working on master branch).

Let me know when you have downloaded the file so I can delete it.

teejusb commented:
Finally getting around to trying this but currently when testing on Mac it seems like my game boots fine, but seemingly because it doesn't get loaded.

I put the zip into the Packages folder and booted the game. Game boots fine but still just shows the default 3 songs. I tried renaming it to .smzip and clicked on it to open it with ITGmania and it did say success installing but still no real difference.

This is with ITGmania 0.7.0 even without this PR applied.

augustg commented:
Checking out the beta branch I get the following log messages:
```
Log starting 2024-03-17 20:00:53
VFS: found /Packages/LargeVideoFile.zip
"zip", "/Packages/LargeVideoFile.zip", "/"
/////////////////////////////////////////
WARNING: C:\Projects\itgmania\src\RageFileDriverZip.cpp:134: Couldn't open /Packages/LargeVideoFile.zip: couldn't find end of central directory record
/////////////////////////////////////////
```
So step 2 in my previous comment was incorrect, the game starts but ignores the package.

On large-file-support2 it works (on my machine...).

What does the beginning of your log look like (where it mounts all filesystems) on this branch?

augustg committed:
Use int64_t in FileDB

sukibaby commented:
If we are still looking to get this change in, I believe the File struct's data types need updating as well. I don't believe I can add a commit to this PR, but I can get that change published. I don't think this will fully work until that is done

augustg commented:
I'll look into it.

sukibaby commented:
If you want me to handle that in a separate commit just lmk.

augustg committed:
Merge from beta

augustg committed:
static_cast to target type

augustg commented on src/RageFileDriverDirectHelpers.cpp:
```diff
@@ -205,7 +205,7 @@ void DirectFilenameDB::CacheFile( const RString &sPath )
 	{
 		f.dir = S_ISDIR(st.st_mode);
-		f.size = static_cast<int>(st.st_size);
+		f.size = static_cast<int64_t>(st.st_size);
```
So after having a look I don't see what needs changing in the File struct (RageUtil_FileDB.h). Am I missing something?

This ended up being the thing I found that needed attention. #348 made a change a few lines above constructing a 64-bit file size value, but then proceeded to store it in the int `size` field on the `File` struct.
With this PR the `size` field is now an int64_t so that should be resolved, but the non-win32 code path still (naturally) had a cast to int. `st.st_size` is an off_t and should virtually always be 64-bit.

sukibaby replied:
As you mentioned, the File struct stores size and hash in an `int`, so that needs to be updated to a 64 bit integer, because the shared code between ITGm & SM5.1 / older Stepmania is very prone to undefined/unexpected/inconsistent behavior, so I'd prefer to explicitly static_cast to int64 for hash and size here in `RageFileDriverDirectHelpers.cpp` to make sure nothing gets truncated as it's passed around the engine, and then update the struct in `RageUtil_FileDB.h`. And then I think we should be good here.

sukibaby commented:
Btw. One last thing to note, I just read over the original comment on this PR for the first time in a long time.

I noted the original reason for supporting 2GB+ was for bulk processing songs for a consistent sync offset in a `smzip`. With "Default Sync Bias" being implemented into the engine now,  you can more easily handle packs with differing sync offset values by adding a `Pack.ini` to the folder of any pack that doesn't match up with the content your game is synced to. 

I would still like to see this change merged eventually. I would approve of this change if the File struct updated to explicitly use 64 bit wide data types for `size` and `hash`, and the change goes through testing across OS's to make sure no unexpected behavior arises from altering the File struct.

With this said,  I am left wondering if there is no real use case for supporting 2GB+ files now that we can mark bias by pack with `Pack.ini`. If `Pack.ini` does not line up with your workflow I'd also be interested to know why.

augustg committed:
Change FilenameDB::GetFileSize to int64_t

augustg committed:
Update misc usage of file size to int64_t

augustg committed:
Update ThreadedFileWorker to use int64_t

augustg committed:
Fix hash size

augustg commented:
I found a few more places where int was used to represent size and fixed them. And I also fixed the hash issue by properly shortening the hash to 32 bits as there is no benefit to using 64 bits in this case. Changing to 64 bits is just another rabbit hole of hard to track down changes. After almost 2 years I still found a few more places to change size data types today.

About my use case. My impression is that the community would like eventually get rid of the +9ms bias. So having all packs follow the same (null) syncing paradigm is the endgame. I'm doing it. Pack.ini would a step backwards (for me). 
But ignoring the goals of the community, having packs with different sync is another layer of complexity. And its just not very neat :).

I hope you're happy with the hash solution — it saves us a lot of headaches. And there is nothing stopping us from switching to 64 bits in future.

teejusb commented:
There are seemingly conflicts that need to be resolved. I can review after.

augustg committed:
Merge branch 'beta' into large-file-support2

augustg commented:
The branch was plenty behind beta, but no conflicts as far as I can see. 
Unless you are referring to rebase conflicts. When completing/merging the PR, do a squash instead. My commit messages don't add much value.

augustg committed:
Fix POSIX lseek name

augustg committed:
Fix std::min compile error on mac

augustg committed:
Include cstdint in RageFileManager.h

augustg committed:
Merge branch 'beta' into large-file-support2

augustg commented:
I just merged the latest beta into this branch. Confirmed locally that it builds on ubuntu and windows.

Could someone trigger a CI build?
